storage/imx6ull-flash/flashsrv.c

```diff
@@ -1,637 +1,372 @@
 
-	return writesz;
+	return strg->dev->mtd->ops->read(strg, strg->start + offs, data, size);
 }
 
 
-static int flashsrv_devWriteMeta(flash_i_devctl_t *idevctl, char* data)
+static ssize_t flashsrv_write(oid_t *oid, size_t offs, char *data, size_t size)
 {
-	flashdrv_dma_t *dma;
-	int i, err;
-	char *databuf;
-	size_t partoff;
-	size_t size = idevctl->write.size;
-	size_t writesz = idevctl->write.size;
+	storage_t *strg = storage_get(oid->id);
 
-	TRACE("META write off: %d, size: %d, ptr: %p", idevctl->write.address, size, data);
+	TRACE("Write off: %d, size: %d, ptr: %p", offs, size, data);
 
-	if (size % flashsrv_common.info.writesz)
+	/* TODO: erase memory area for data == NULL ??? feature or bug */
+	if (strg == NULL || strg->dev == NULL || strg->dev->ctx == NULL || strg->dev->mtd == NULL || strg->dev->mtd->ops == NULL
+			|| strg->dev->mtd->ops->write == NULL || (offs + size) > strg->size || data == NULL) {
 		return -EINVAL;
-
-	if (idevctl->write.address % flashsrv_common.info.writesz)
-		return -EINVAL;
-
-	if ((err = flashsrv_partoff(idevctl->write.oid.id, 0, idevctl->write.address, idevctl->write.size, &partoff, NULL)) < 0)
-		return err;
-
-	idevctl->write.address += partoff;
-
-	dma = flashsrv_common.dma;
-	databuf = flashsrv_common.databuf;
-
-	memcpy(databuf, data, flashsrv_common.info.writesz);
-	for (i = 0; size; i++) {
-		TRACE("  paddr: %u, eb: %u", idevctl->write.address / flashsrv_common.info.writesz, idevctl->write.address / flashsrv_common.info.erasesz);
-		err = flashdrv_write(dma, idevctl->write.address / flashsrv_common.info.writesz + i, NULL, databuf);
-
-		if (err) {
-			LOG_ERROR("write error %d", err);
-			break;
-		}
-		size -= flashsrv_common.info.writesz;
 	}
 
-	writesz -= size;
-
-	return writesz;
+	return strg->dev->mtd->ops->write(strg, strg->start + offs, data, size);
 }
 
 
-static int flashsrv_devReadRaw(flash_i_devctl_t *idevctl, char *data)
+static int flashsrv_fileAttrGet(oid_t *oid, int type, long long *attr)
 {
-	flashdrv_dma_t *dma;
-	char *databuf;
-	size_t rp, partoff, totalBytes = 0;
-	const unsigned int raw_page_size = flashsrv_common.info.writesz + flashsrv_common.info.metasz;
-	size_t size = idevctl->readraw.size;
-	size_t offset = idevctl->readraw.address;
-	int err = EOK;
+	storage_t *strg = storage_get(oid->id);
 
-	if ((size % raw_page_size) || (offset % raw_page_size))
+	if (strg == NULL) {
 		return -EINVAL;
+	}
 
-	if ((err = flashsrv_partoff(idevctl->readraw.oid.id, 1, offset, size, &partoff, NULL)) < 0) /* raw = 1 */
-		return err;
-
-	offset += partoff;
-
-	dma = flashsrv_common.dma;
-	databuf = flashsrv_common.databuf;
-	rp = offset / raw_page_size;
-
-	while (size) {
-		err = flashdrv_readraw(dma, rp, databuf, raw_page_size);
-		memcpy(data, databuf, raw_page_size);
+	switch (type) {
+		case atSize:
+			*attr = (off_t)strg->size;
+			break;
 
-		if (err < 0) {
-			LOG_ERROR("error in readraw(): %d", err);
-			err = -EIO;
+		case atDev:
+			*attr = (off_t)strg->start;
```

Comment:
I am pretty sure that's not what atDev is for

Reply:
It's used in `psd` - not sure why: [link](https://github.com/phoenix-rtos/phoenix-rtos-utils/blob/a46e92d6e5002941b541d14b019d6139f2c18cc5/psd/imx6ull/psd.c#L177)
We should leave it for now for `psd` to work.